crates/biome_js_analyze/src/lint/nursery/no_label_without_control.rs

```diff
@@ -0,0 +1,290 @@
+    }
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoLabelWithoutControlOptions {
+    pub label_components: Vec<String>,
+    pub label_attributes: Vec<String>,
+    pub control_components: Vec<String>,
+}
+
+pub struct NoLabelWithoutControlState {
+    pub has_text_content: bool,
+    pub has_control_association: bool,
+}
+
+impl Rule for NoLabelWithoutControl {
+    type Query = Ast<AnyJsxTag>;
+    type State = NoLabelWithoutControlState;
+    type Signals = Option<Self::State>;
+    type Options = NoLabelWithoutControlOptions;
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let options = ctx.options();
+        let label_attributes = get_option(&options.label_attributes, &["aria-label", "alt"]);
+        let label_components = get_option(&options.label_components, &["label"]);
+        let control_components = get_option(
+            &options.control_components,
+            &["input", "meter", "output", "progress", "select", "textarea"],
+        );
+        let element_name = get_element_name(node)?;
+        let is_allowed_element = label_components.contains(&element_name);
+
+        if !is_allowed_element {
+            return None;
+        }
+
+        let has_text_content = has_accessible_label(node, &label_attributes);
+        let has_control_association =
+            has_for_attribute(node)? || has_nested_control(node, &control_components);
+
+        if has_text_content && has_control_association {
+            return None;
+        }
+
+        Some(NoLabelWithoutControlState {
+            has_text_content,
+            has_control_association,
+        })
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+        let mut diagnostic = RuleDiagnostic::new(
+            rule_category!(),
+            node.range(),
+            markup! {
+                "A form label must be associated with a control."
+            },
+        );
+
+        if !state.has_text_content {
+            diagnostic = diagnostic.note(
+                markup! { "Consider adding an accessible text content to the label element." },
+            );
+        }
+
+        if !state.has_control_association {
+            diagnostic = diagnostic.note(
+                markup! { "Consider adding a `for` or `htmlFor` attribute to the label element or moving the control element to inside the label element." },
+            );
+        }
+
+        Some(diagnostic)
+    }
+}
+
+/// Returns an array option merged with the default value
+fn get_option(value: &[String], default_value: &[&str]) -> Vec<String> {
+    value
+        .iter()
+        .cloned()
+        .chain(default_value.iter().map(|value| (*value).to_string()))
+        .collect()
+}
```

Comment:
I am not sure this is a good strategy. While it makes it easier code-wise, we are allocating strings, and it isn't needed. We could easily create a static array. Sure, we would need to do two checks (array and options), but we don't crate strings at runtime.

Reply:
changed it to check both the default array and custom options manually